frontend/cs/scs/builder.go

```diff
@@ -686,3 +687,57 @@ func (builder *builder) ToCanonicalVariable(v frontend.Variable) frontend.Canoni
 		return term
 	}
 }
+
+// GetWireGates returns the pairs (gateId, wireLocation) for the given wires in
+// the compiled constraint system:
+//   - gateId is the index of the gate in the constraint system.
+//   - wireLocation is the location of the wire in the gate (0=xA or 1=xB).
+//
+// If the argument addMissing is true, then the function will add a new
+// instruction for each wire that is not found in the constraint system. This
+// may happen when getting the gate for a witness which is not used in gates.
+// Otherwise, when addMissing is false, the function returns an error if a wire
+// is not found in the constraint system.
+//
+// The order of the returned pairs is not the same as for the given arguments.
+func (builder *builder) GetWireGates(wires []frontend.Variable, addMissing bool) ([][2]int, error) {
+	// construct a lookup table table for later quick access when iterating over instructions
+	lookup := make(map[int]struct{})
+	for _, w := range wires {
+		ww, ok := w.(expr.Term)
+		if !ok {
+			panic("input wire is not a Term")
+		}
+		lookup[ww.WireID()] = struct{}{}
+	}
+	res := make([][2]int, 0, len(wires))
+	iterator := builder.cs.GetSparseR1CIterator()
+	for c, gate := iterator.Next(), 0; c != nil; c, gate = iterator.Next(), gate+1 {
+		if _, ok := lookup[int(c.XA)]; ok {
+			res = append(res, [2]int{gate, 0})
+			delete(lookup, int(c.XA))
+			continue
+		}
+		if _, ok := lookup[int(c.XB)]; ok {
+			res = append(res, [2]int{gate, 1})
+			delete(lookup, int(c.XB))
+			continue
+		}
+	}
+	if addMissing {
```

Comment:
if this can happen iff the wires are the witness, shouldn't we sanity check the wireIDs against the range of wireIDs for the witness ?

Reply:
I think it makes sense to check wireIDs agains wireIDs of the witness. I think otherwise it is a bug in the builder itself when we have created a non-witness wire which doesn't appear in any constraint. I think the best approach would be to panic.

Reply:
Added sanity check.